Different handling of unreachable code and default fallthrough in switch block

**Version Used**: 
Visual Studio 2019 RC (csc.exe 3.0.19.12206 (ec366687))

**Steps to Reproduce**:
Compile this code (I used `csc.exe Class1.cs -target:library -langversion:7.3`):

    int Foo(bool a)
    {
        switch (a)
        {
            case true:
                return 1;
            case false:
                return 2;
            default:
                return 3;
        }
    }

**Expected Behavior**:
No warnings.

**Actual Behavior**:
You'll get warning CS0162: Unreachable code detected. This wasn't the case with Visual Studio 2017 (csc.exe 2.10.0.0 (b9fb1610)). This warning is of course promoted to an error if warnings as errors is enabled, which is how I found it. Is this a problem? I realize that analysis may have improved.

However, 2019 then suggests replacing it with:

    int Foo(bool a)
    {
        switch (a)
        {
            case true:
                return 1;
            case false:
                return 2;
            default:
        }
    }

Which now compiles without warnings. This code *fails* with Visual Studio 2017:

    CS8070: Control cannot fall out of switch from final case label ('default')
    CS0161: 'Class1.Foo(bool)': not all code paths return a value

But I am using langversion 7.3 for both. Shouldn't the same language version mean that it should compile with the same result?

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_Switch.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_Switch.cs
@@ -30,7 +30,8 @@ namespace Microsoft.CodeAnalysis.CSharp
             {
                 foreach (var label in section.SwitchLabels)
                 {
-                    if (reachableLabels.Contains(label.Label) || label.HasErrors)
+                    if (reachableLabels.Contains(label.Label) || label.HasErrors ||
+                        label == node.DefaultLabel && node.Expression.ConstantValue == null && IsTraditionalSwitch(node))
                     {
                         SetState(initialState.Clone());
                     }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -1155,7 +1155,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 this.StateWhenFalse[mainSlot] = whenFalse;
             }
 
-            if (whenTrue.IsNotNull()|| whenFalse.IsNotNull())
+            if (whenTrue.IsNotNull() || whenFalse.IsNotNull())
             {
                 var slotBuilder = ArrayBuilder<int>.GetInstance();
                 GetSlotsToMarkAsNotNullable(expression, slotBuilder);

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/SwitchTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/SwitchTests.cs
@@ -2918,6 +2918,35 @@ public class TestClass
                 );
         }
 
+        [Fact]
+        [WorkItem(33783, "https://github.com/dotnet/roslyn/issues/33783")]
+        public void UnreachableDefaultInBoolSwitch()
+        {
+            var text = @"
+public class TestClass
+{
+    public static void Main()
+    {
+        bool b = false;
+        switch (b)
+        {
+            case true:
+                break;
+            case false:
+                break;
+            default:
+                break; //1
+        }
+    }
+}";
+            CreateCompilation(text, parseOptions: TestOptions.Regular6).VerifyDiagnostics();
+            CreateCompilation(text, parseOptions: TestOptions.Regular7_3).VerifyDiagnostics();
+            CreateCompilation(text).VerifyDiagnostics(
+                // (14,17): warning CS0162: Unreachable code detected
+                //                 break; //1
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "break").WithLocation(14, 17));
+        }
+
         #endregion
     }
 }

